### PR TITLE
Add .step files hadneler, create base NURBS class, change tests *NOT READY TO MERGE*

### PR DESCRIPTION
**Pull request by maciekgroch**
First of all,
=
I started with implementing functionality for STEP files. As it is nearly the same as for IGES it was decided to create a base class for the two -- `nurbshandler` which is an abstract class from which both `igeshandler` and `stephandler` inherit.

Implementing step functionality was the main goal while developing the given code. However, during the process I encountered multiple issues, which I decided to handle. As it can be seen refactoring got a bit out of control 🐑 .

NURBS-, STEP-, IGESHANDLERS
=
1.It should be reviewed if the code which was used for *.igs is interchangeable  for *.step. It concerns `write` and `parse` methods.
2. Reading files from disk was repeated in multiple places. Following the 'do not repeat yourself' rule I decided to create a `@classmethod load_shape_from_file` to avoid it. Moreover, in multiple applications I wanted to modify the shape before parsing it and it is more convenient to have a method for that.
3. The same was done with `@classmethod write_shape_to_file` 
4. It was decided that the loaded shape should be stored inside the object (`shape`). It is always the same shape which was loaded that is later written to a file. I see no point in reading it multiple times. It only makes code prone to errors.

HANDLERS GENERALLY
=
1. List of possible extensions is constant. It was never changed inside the code (and I strongly believe it should not be). As a result it was moved from a variable to a class constant. (I don't know if it is well described in the docstrings -- please review).
2. Method `_check_infile_instantiation` in `filehandler` is no longer static. Since it is always `self.infile` that is checked, IMHO there is no point in keeping it static -- let's just check `self.infile` inside.
3. Method `_check_extension` is now a `classmethod`. Since extensions are class constants it was decided to refactor the method this way.

TESTS
=
I tried to clean tests as well.
1. I got rid of classic `assert x == y`. Used `unittest` asserts.
2. Instead of simply `os.remove(x)` at the end of the method `addCleanup` is added  in order to make sure that tmp file are always deleted. 

THINGS I AM NOT SURE ABOUT
=
Tests for the `stephandler` class do not pass. This is my concern: should the positions of the U,V poles be the same as for IGES? IMHO not. I provide you with a small comparison I made:

unnamed

It seems that the points are positioned correctly. Nevertheless, there is no point in creating a test where I simply put results of the tested method. What do you recommend as a way of testing it if step files are parsed correctly?

Things that I would like to change, however, I decided to leave it for later. 
=
1. `write` method breaks the 'single responsibility rule' -- it does more than writing the shape to the file. Moreover, sometimes it is more convenient to get a shape after modification (saving it just to read it a moment later is frustrating). 
2. `parse` method should also take `TopoDS_Shape` not only a file. The argumentation is the same as in the previous case. 

Lastly,
=
I know there is a lot of code to review. Nonetheless, I believe that it will boost the usability of the package. Enjoy 🔫

### CODE REVIEW

**Comment by mtezzele** [comment]
Dear Maciek, this week I am abroad for a conference. I will review your contribution but it will take time. I am sorry. Thanks again for your work!

**Comment by mtezzele** [comment]
Please make sure that all the tutorials still work.

**Comment by mtezzele** [comment]
@maciekgroch any news?

**Comment by maciekgroch** [comment]
I have super busy time at work, sorry for delay. I promise to work on that
during the weekend!

17 lis 2016 15:11 "Marco Tezzele" <notifications@github.com> napisał(a):

File: tests/test_igeshandler.py [hunk 1]
```diff
@@ -77,6 +64,12 @@ def test_iges_parse_coords_2(self):
 	def test_iges_parse_coords_3(self):
 		iges_handler = ih.IgesHandler()
 		mesh_points = iges_handler.parse('tests/test_datasets/test_pipe.iges')
+		print np.max(mesh_points[:, 0])
```
Comment (maciekgroch) [inline tests/test_igeshandler.py]: done.

File: tests/test_package.py [hunk 2]
```diff
@@ -50,6 +58,7 @@ def test_modules_name(self):
 		f_aux = []
 		for (__, __, filenames) in walk('pygem'):
 			f_aux.extend(filenames)
+			print filenames
```
Comment (maciekgroch) [inline tests/test_package.py]: Done.

File: pygem/stephandler.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+		This class method saves the `shape` to the file `filename`.
+
+		:param: TopoDS_Shape shape: loaded shape
+		:param string filename: name of the input file.
+			It should have proper extension (.step or .stp)
+		"""
+		cls._check_filename_type(filename)
+		cls._check_extension(filename)
+		step_writer = STEPControl_Writer()
+		Interface_Static_SetCVal("write.step.schema", "AP203")
```
Comment (maciekgroch) [inline pygem/stephandler.py]: done.

**Comment by maciekgroch** [comment]
@mtezzele, All the issues you pointed out were fixed. In addition to that, I found some other problems, which I did not notice before (e.g. STEP handling in GUI). Tutorials are ok -- they were run, and no problems occured.

**Comment by mtezzele** [comment]
@maciekgroch it seems fine to me. Please fix the failing tests, so we can close merge this PR.

**Comment by maciekgroch** [comment]
I believe I fixed all the issues but the following ones: `test_iges_write_comparison_iges` and `test_iges_write_comparison_igs`. However, the last two do not pass even on the master branch.

**Pull request merged**